File: DSA-lab/modules/02-arrays-strings/projects/01-basic/growth_experiment.py
"""Visualize capacity growth and amortized append costs."""

from __future__ import annotations

import argparse
import csv
from dataclasses import dataclass
from pathlib import Path
from typing import Sequence

from dynamic_array import DynamicArray

DEFAULT_APPENDS = 64
DEFAULT_OUTPUT_DIR = Path(__file__).resolve().parent / "results"


@dataclass(frozen=True)
class AppendMeasurement:
    """Operation-model data for one append."""

    append_number: int
    length: int
    capacity: int
    copied_elements: int
    actual_cost: int
    cumulative_average_cost: float


def run_experiment(
    append_count: int,
    initial_capacity: int = 1,
) -> list[AppendMeasurement]:
    """Append integers and record deterministic operation costs."""
    if append_count < 1:
        raise ValueError("append_count must be positive")
    if initial_capacity < 1:
        raise ValueError("initial_capacity must be at least 1")

    array: DynamicArray[int] = DynamicArray(initial_capacity)
    measurements = []
    cumulative_cost = 0

    for value in range(append_count):
        event_count_before = len(array.growth_events)
        array.append(value)

        copied_elements = 0
        if len(array.growth_events) > event_count_before:
            copied_elements = array.growth_events[-1].copied_elements

        actual_cost = 1 + copied_elements
        cumulative_cost += actual_cost
        append_number = value + 1
        measurements.append(
            AppendMeasurement(
                append_number=append_number,
                length=len(array),
                capacity=array.capacity,
                copied_elements=copied_elements,
                actual_cost=actual_cost,
                cumulative_average_cost=(
                    cumulative_cost / append_number
                ),
            )
        )

    return measurements


def write_csv(
    measurements: Sequence[AppendMeasurement],
    destination: Path,
) -> None:
    """Write experiment measurements to CSV."""
    destination.parent.mkdir(parents=True, exist_ok=True)
    with destination.open("w", newline="", encoding="utf-8") as csv_file:
        writer = csv.writer(csv_file)
        writer.writerow(
            [
                "append_number",
                "length",
                "capacity",
                "copied_elements",
                "actual_cost",
                "cumulative_average_cost",
            ]
        )
        for point in measurements:
            writer.writerow(
                [
                    point.append_number,
                    point.length,
                    point.capacity,
                    point.copied_elements,
                    point.actual_cost,
                    f"{point.cumulative_average_cost:.8f}",
                ]
            )


def create_plot(
    measurements: Sequence[AppendMeasurement],
    destination: Path,
) -> None:
    """Create capacity and append-cost plots."""
    import matplotlib

    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    destination.parent.mkdir(parents=True, exist_ok=True)
    append_numbers = [point.append_number for point in measurements]

    figure, (capacity_axis, cost_axis) = plt.subplots(
        2,
        1,
        figsize=(11, 8),
        sharex=True,
    )
    capacity_axis.step(
        append_numbers,
        [point.capacity for point in measurements],
        where="post",
        linewidth=2,
        color="tab:blue",
        label="Buffer capacity",
    )
    capacity_axis.plot(
        append_numbers,
        [point.length for point in measurements],
        linestyle="--",
        color="tab:gray",
        label="Logical length",
    )
    capacity_axis.set_ylabel("Positions")
    capacity_axis.set_title("Geometric buffer growth")
    capacity_axis.grid(True, linestyle=":", alpha=0.65)
    capacity_axis.legend()

    cost_axis.vlines(
        append_numbers,
        0,
        [point.actual_cost for point in measurements],
        color="tab:red",
        alpha=0.55,
        label="Actual cost per append",
    )
    cost_axis.plot(
        append_numbers,
        [point.cumulative_average_cost for point in measurements],
        color="tab:green",
        linewidth=2,
        label="Cumulative average cost",
    )
    cost_axis.axhline(
        3,
        color="tab:purple",
        linestyle=":",
        label="Upper amortized bound (< 3)",
    )
    cost_axis.set_xlabel("Append number")
    cost_axis.set_ylabel("Abstract operation units")
    cost_axis.set_title("Rare resize spikes, bounded amortized average")
    cost_axis.grid(True, linestyle=":", alpha=0.65)
    cost_axis.legend()

    figure.suptitle("DynamicArray: capacity and amortized append cost")
    figure.tight_layout()
    figure.savefig(destination, dpi=160)
    plt.close(figure)


def print_summary(measurements: Sequence[AppendMeasurement]) -> None:
    """Print resize points and final amortized cost."""
    resize_points = [
        point for point in measurements if point.copied_elements > 0
    ]

    print("Resize events")
    print("=" * 58)
    for point in resize_points:
        print(
            f"append={point.append_number:4d}, "
            f"capacity={point.capacity:4d}, "
            f"copied={point.copied_elements:4d}, "
            f"actual_cost={point.actual_cost:4d}"
        )

    final = measurements[-1]
    total_copies = sum(point.copied_elements for point in measurements)
    print("\nSummary")
    print(f"  appends:                 {final.append_number}")
    print(f"  final capacity:          {final.capacity}")
    print(f"  total resize copies:     {total_copies}")
    print(
        f"  cumulative average cost: "
        f"{final.cumulative_average_cost:.4f}"
    )


def parse_arguments() -> argparse.Namespace:
    """Parse command-line options."""
    parser = argparse.ArgumentParser(
        description="Visualize dynamic-array growth and amortized append cost."
    )
    parser.add_argument("--appends", type=int, default=DEFAULT_APPENDS)
    parser.add_argument("--initial-capacity", type=int, default=1)
    parser.add_argument("--output-dir", type=Path, default=DEFAULT_OUTPUT_DIR)
    return parser.parse_args()


def main() -> None:
    """Run the experiment and persist its artifacts."""
    arguments = parse_arguments()
    try:
        measurements = run_experiment(
            arguments.appends,
            arguments.initial_capacity,
        )
    except ValueError as error:
        raise SystemExit(f"Configuration error: {error}") from error

    csv_path = arguments.output_dir / "growth_log.csv"
    plot_path = arguments.output_dir / "capacity_and_costs.png"
    write_csv(measurements, csv_path)
    create_plot(measurements, plot_path)
    print_summary(measurements)

    print("\nArtifacts")
    print(f"  CSV:  {csv_path}")
    print(f"  Plot: {plot_path}")


if __name__ == "__main__":
    main()

File: DSA-lab-java/modules/02-arrays-strings/projects/01-basic/src/main/java/lab/dsa/module02/basic/GrowthExperiment.java
package lab.dsa.module02.basic;

import java.awt.Color;
import java.awt.Graphics2D;
import java.awt.image.BufferedImage;
import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import javax.imageio.ImageIO;

/** Visualizes capacity growth and amortized append costs. */
public final class GrowthExperiment {
    private GrowthExperiment() {
    }

    public record AppendMeasurement(int appendNumber, int size, int capacity,
                                    int copiedElements, int actualCost,
                                    double cumulativeAverageCost) {
    }

    public static List<AppendMeasurement> runExperiment(int appendCount, int initialCapacity) {
        if (appendCount < 1) throw new IllegalArgumentException("Append count must be positive");
        if (initialCapacity < 1) throw new IllegalArgumentException("Initial capacity must be at least 1");
        DynamicArray<Integer> array = new DynamicArray<>(initialCapacity);
        List<AppendMeasurement> measurements = new ArrayList<>();
        long cumulativeCost = 0;
        for (int value = 0; value < appendCount; value++) {
            int eventCountBefore = array.growthEvents().size();
            array.append(value);
            int copied = array.growthEvents().size() > eventCountBefore
                    ? array.growthEvents().getLast().copiedElements() : 0;
            int cost = 1 + copied;
            cumulativeCost += cost;
            int appendNumber = value + 1;
            measurements.add(new AppendMeasurement(appendNumber, array.size(), array.capacity(),
                    copied, cost, (double) cumulativeCost / appendNumber));
        }
        return measurements;
    }

    public static void writeCsv(List<AppendMeasurement> measurements, Path destination) throws IOException {
        if (destination.getParent() != null) Files.createDirectories(destination.getParent());
        try (BufferedWriter writer = Files.newBufferedWriter(destination, StandardCharsets.UTF_8)) {
            writer.write("append_number,length,capacity,copied_elements,actual_cost,cumulative_average_cost");
            writer.newLine();
            for (AppendMeasurement point : measurements) {
                writer.write(String.format(Locale.ROOT, "%d,%d,%d,%d,%d,%.8f", point.appendNumber(),
                        point.size(), point.capacity(), point.copiedElements(), point.actualCost(),
                        point.cumulativeAverageCost()));
                writer.newLine();
            }
        }
    }

    public static void createPlot(List<AppendMeasurement> measurements, Path destination) throws IOException {
        int width = 1_100;
        int height = 800;
        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D graphics = image.createGraphics();
        graphics.setColor(Color.WHITE); graphics.fillRect(0, 0, width, height);
        graphics.setColor(Color.BLACK); graphics.drawString("DynamicArray: capacity and amortized append cost", 400, 25);
        drawSeries(graphics, measurements, 70, 70, 950, 280, true);
        drawSeries(graphics, measurements, 70, 440, 950, 280, false);
        graphics.dispose();
        if (destination.getParent() != null) Files.createDirectories(destination.getParent());
        ImageIO.write(image, "png", destination.toFile());
    }

    private static void drawSeries(Graphics2D graphics, List<AppendMeasurement> points,
                                   int x, int y, int width, int height, boolean capacity) {
        graphics.setColor(Color.BLACK); graphics.drawRect(x, y, width, height);
        graphics.drawString(capacity ? "Geometric buffer growth" : "Append cost and cumulative average", x + 330, y - 12);
        double maximum = capacity
                ? points.stream().mapToDouble(AppendMeasurement::capacity).max().orElseThrow()
                : points.stream().mapToDouble(AppendMeasurement::actualCost).max().orElseThrow();
        int previousX = -1; int previousY = -1;
        graphics.setColor(capacity ? Color.BLUE : Color.RED);
        for (int index = 0; index < points.size(); index++) {
            AppendMeasurement point = points.get(index);
            double value = capacity ? point.capacity() : point.actualCost();
            int currentX = x + index * width / Math.max(1, points.size() - 1);
            int currentY = y + height - (int) (value / maximum * height);
            if (previousX >= 0) graphics.drawLine(previousX, previousY, currentX, currentY);
            previousX = currentX; previousY = currentY;
        }
        if (!capacity) {
            graphics.setColor(new Color(0, 140, 0)); previousX = -1; previousY = -1;
            for (int index = 0; index < points.size(); index++) {
                int currentX = x + index * width / Math.max(1, points.size() - 1);
                int currentY = y + height - (int) (points.get(index).cumulativeAverageCost() / maximum * height);
                if (previousX >= 0) graphics.drawLine(previousX, previousY, currentX, currentY);
                previousX = currentX; previousY = currentY;
            }
        }
    }

    public static void printSummary(List<AppendMeasurement> measurements) {
        System.out.println("Resize events\n" + "=".repeat(58));
        measurements.stream().filter(point -> point.copiedElements() > 0).forEach(point ->
                System.out.printf("append=%4d, capacity=%4d, copied=%4d, actual_cost=%4d%n",
                        point.appendNumber(), point.capacity(), point.copiedElements(), point.actualCost()));
        AppendMeasurement last = measurements.getLast();
        int totalCopies = measurements.stream().mapToInt(AppendMeasurement::copiedElements).sum();
        System.out.printf(Locale.ROOT, "\nSummary%n  appends: %d%n  final capacity: %d%n  total resize copies: %d%n  cumulative average cost: %.4f%n",
                last.appendNumber(), last.capacity(), totalCopies, last.cumulativeAverageCost());
    }

    public static void main(String[] arguments) throws IOException {
        int appendCount = 64; int initialCapacity = 1; Path outputDirectory = Paths.get("results");
        for (int index = 0; index < arguments.length; index++) {
            switch (arguments[index]) {
                case "--appends" -> appendCount = Integer.parseInt(arguments[++index]);
                case "--initial-capacity" -> initialCapacity = Integer.parseInt(arguments[++index]);
                case "--output-dir" -> outputDirectory = Paths.get(arguments[++index]);
                default -> throw new IllegalArgumentException("Unknown option: " + arguments[index]);
            }
        }
        List<AppendMeasurement> measurements = runExperiment(appendCount, initialCapacity);
        Path csvPath = outputDirectory.resolve("growth_log.csv");
        Path plotPath = outputDirectory.resolve("capacity_and_costs.png");
        writeCsv(measurements, csvPath); createPlot(measurements, plotPath); printSummary(measurements);
        System.out.println("\nArtifacts\n  CSV:  " + csvPath + "\n  Plot: " + plotPath);
    }
}
